modules/renter/repair.go

```diff
@@ -185,6 +202,39 @@ func (f *file) expiringContracts(height types.BlockHeight) []fileContract {
 	return expiring
 }
 
+// isOffline decides whether a host should be considered offline, based on its
+// scan metrics.
+func isOffline(host modules.HostDBEntry) bool {
+	// consider a host offline if:
+	// 1) The host has been scanned at least three times, and
+	// 2) The three most recent scans have all failed, and
+	// 3) The time between the most recent scan and the last successful scan
+	//    (or first scan) is at least uptimeWindow
+	numScans := len(host.ScanHistory)
+	if numScans < uptimeMinScans {
+		// not enough data to make a fair judgment
+		return false
+	}
+	// NOTE: ScanHistory is ordered from oldest-newest
+	recent := host.ScanHistory[numScans-uptimeMinScans:]
+	for _, scan := range recent {
+		if scan.Success {
+			// one of the scans succeeded
+			return false
+		}
+	}
+	// initialize window bounds
+	windowStart, windowEnd := host.ScanHistory[0].Timestamp, host.ScanHistory[numScans-1].Timestamp
+	// iterate from newest-oldest, seeking to last successful scan
+	for i := numScans - 1; i >= 0; i-- {
+		if scan := host.ScanHistory[i]; scan.Success {
+			windowStart = scan.Timestamp
+			break
+		}
+	}
+	return windowEnd.Sub(windowStart) >= uptimeWindow
+}
```

Comment:
hmm.

Given how this is used, I'm going to  say that this should be established as an exported method of the contract. "Is this contract still useful at all?" seems like a question that the contractor should be answering, given that it's the one collecting the metrics on contracts and whatnot.

The renter will have separate sorts of 'should I use this contract right now' functions, (e.g. 'no' if it has failed to perform an upload in the past 2 hours or if it has been unbearably slow).

With regards to my previous position on duplicating this code, I had assumed we were talking about duplicating it between the hostdb and contractor, where I feel it is being used in sufficiently different ways to justify separate implementations.